enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -686,40 +693,56 @@ func (m *Mmap) initMap() error {
 }
 
 func (m *Mmap) ReadAt(p []byte, off int64) (n int, err error) {
-	if !m.mapped {
+	if !m.safeReadMapped() {
 		if err := m.initMap(); err != nil {
 			return 0, err
 		}
 	}
-	if err := checkBounds("read", int64(len(m.data)), p, off); err != nil {
+	dataLen, err := m.SizeBytes()
+	if err != nil {
+		return 0, err
+	}
+	if err := checkBounds("read", dataLen, p, off); err != nil {
 		return 0, err
 	}
+	m.mu.RLock()
+	defer m.mu.RUnlock()
 	copy(p, m.data[int(off):int(off)+len(p)])
 	return len(p), nil
 }
 
 func (m *Mmap) WriteAt(p []byte, off int64) (n int, err error) {
-	if !m.mapped {
+	if !m.safeReadMapped() {
 		if err := m.initMap(); err != nil {
 			return 0, err
 		}
 	}
-	if err := checkBounds("write", int64(len(m.data)), p, off); err != nil {
+	dataLen, err := m.SizeBytes()
+	if err != nil {
+		return 0, err
+	}
+	if err := checkBounds("write", dataLen, p, off); err != nil {
 		return 0, err
 	}
+	m.mu.Lock()
 	m.lazyDigest = nil
 	copy(m.data[int(off):int(off)+len(p)], p)
+	m.mu.Unlock()
```

Comment:
nit: defer the unlock

```suggestion
	m.mu.Lock()
	defer m.mu.Unlock()
	m.lazyDigest = nil
	copy(m.data[int(off):int(off)+len(p)], p)
```